Refactor error msg stack handling, add TORCH_RETHROW

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37101 Refactor error msg stack handling, add TORCH_RETHROW**
* #37094 Make msg() and msg_with_backtrace() private

Fixes #36954.

The basic concept is to streamline the process of rethrowing
c10::Error with extra error information.  This is in a few
steps:

- I completely remodeled the Error data type and the internal
  invariants.  Instead of manually adding in newlines, the
  message stack formatting process is responsible for inserting
  newlines and spacing as necessary.  Call sites are then
  modified to respect the new API model.
- TORCH_RETHROW macro is added, which adds context to an error
  message and then rethrows it.

New internal assert failure looks like:

```
0 INTERNAL ASSERT FAILED at ../c10/test/util/exception_test.cpp:64, please report a bug to PyTorch.
Exception raised from TestBody at ../c10/test/util/exception_test.cpp:64 (most recent call first):
frame #0: <unknown function> + 0x6aab9 (0x7ff611d3aab9 in /data/users/ezyang/pytorch-tmp/build/lib/libc10.so)
frame #1: ...
```

Error message with context looks like:

```
This is an error
  This is context 1
  This is context 2
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21202891](https://our.internmc.facebook.com/intern/diff/D21202891)

ezyang committed:
Refactor error msg stack handling, add TORCH_RETHROW

Fixes #36954.

The basic concept is to streamline the process of rethrowing
c10::Error with extra error information.  This is in a few
steps:

- I completely remodeled the Error data type and the internal
  invariants.  Instead of manually adding in newlines, the
  message stack formatting process is responsible for inserting
  newlines and spacing as necessary.  Call sites are then
  modified to respect the new API model.
- TORCH_RETHROW macro is added, which adds context to an error
  message and then rethrows it.

New internal assert failure looks like:

```
0 INTERNAL ASSERT FAILED at ../c10/test/util/exception_test.cpp:64, please report a bug to PyTorch.
Exception raised from TestBody at ../c10/test/util/exception_test.cpp:64 (most recent call first):
frame #0: <unknown function> + 0x6aab9 (0x7ff611d3aab9 in /data/users/ezyang/pytorch-tmp/build/lib/libc10.so)
frame #1: ...
```

Error message with context looks like:

```
This is an error
  This is context 1
  This is context 2
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

mruberry commented on c10/util/Exception.cpp:
```diff
@@ -38,29 +38,50 @@ Error::Error(
+    oss << msg_stack_[0] << " (" << msg_stack_[1] << ")";
+  } else {
+    oss << msg_stack_[0];
+    for (size_t i = 1; i < msg_stack_.size(); i++) {
```
nit: auto i = decltype(msg_stack_.size()){1}

ezyang replied:
Come on, is this really what C++ programmers are supposed to write? XD

mruberry replied:
There's some level of hating build warnings where it becomes right?

mruberry commented on c10/util/Exception.cpp:
```diff
@@ -14,8 +15,7 @@ Error::Error(
     const std::string& backtrace,
     const void* caller)
     : msg_stack_{new_msg}, backtrace_(backtrace), caller_(caller) {
```
nit: non-braced initialization?

ezyang replied:
I think the braced initialization here is intentional, they want to create a 1-element vector containing just new_msg

mruberry replied:
I mean, why not backtrace_{backtrace} and caller_{caller}?

ezyang replied:
There's like a chapter in the Modern C++ book about this but I don't remember what it says XD

mruberry reviewed (approved):
LGTM!

bhosmer commented on c10/util/Exception.h:
```diff
@@ -56,13 +66,29 @@ class C10_API Error : public std::exception {
+      const void* caller = nullptr);
+
+  // Add some new context to the message stack (pushing it to the top
+  // of the message stack)
```
Nit: "top" might benefit from some qualification here, since it's actually going to go to the end of the formatted message IIUC

bhosmer commented on c10/util/Logging.cpp:
```diff
@@ -64,7 +64,7 @@ void ThrowEnforceFiniteNotMet(
 // (This must be defined here for access to GetFetchStackTrace)
 Error::Error(SourceLocation source_location, const std::string& msg)
-    : Error(msg, str(" (", source_location, ")\n", (*GetFetchStackTrace())())) {
+    : Error(msg, str("Exception raised from ", source_location, " (most recent call first):\n", (*GetFetchStackTrace())())) {
```
definitely nicer

bhosmer reviewed (approved):
lg, one nontrivial idea inline but def not a blocker

ezyang commented on c10/test/util/exception_test.cpp:
```diff
@@ -2,10 +2,24 @@
+template<class Functor>
+inline void expectThrowsEq(Functor&& functor, const char* expectedMessage) {
+  try {
+    std::forward<Functor>(functor)();
```
I don't know, I copied it from @smessmer's expectThrows helper XD

ezyang committed:
Update on "Refactor error msg stack handling, add TORCH_RETHROW"


Fixes #36954.

The basic concept is to streamline the process of rethrowing
c10::Error with extra error information.  This is in a few
steps:

- I completely remodeled the Error data type and the internal
  invariants.  Instead of manually adding in newlines, the
  message stack formatting process is responsible for inserting
  newlines and spacing as necessary.  Call sites are then
  modified to respect the new API model.
- TORCH_RETHROW macro is added, which adds context to an error
  message and then rethrows it.

New internal assert failure looks like:

```
0 INTERNAL ASSERT FAILED at ../c10/test/util/exception_test.cpp:64, please report a bug to PyTorch.
Exception raised from TestBody at ../c10/test/util/exception_test.cpp:64 (most recent call first):
frame #0: <unknown function> + 0x6aab9 (0x7ff611d3aab9 in /data/users/ezyang/pytorch-tmp/build/lib/libc10.so)
frame #1: ...
```

Error message with context looks like:

```
This is an error
  This is context 1
  This is context 2
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21202891](https://our.internmc.facebook.com/intern/diff/D21202891)

[ghstack-poisoned]

ezyang committed:
Update on "Refactor error msg stack handling, add TORCH_RETHROW"


Fixes #36954.

The basic concept is to streamline the process of rethrowing
c10::Error with extra error information.  This is in a few
steps:

- I completely remodeled the Error data type and the internal
  invariants.  Instead of manually adding in newlines, the
  message stack formatting process is responsible for inserting
  newlines and spacing as necessary.  Call sites are then
  modified to respect the new API model.
- TORCH_RETHROW macro is added, which adds context to an error
  message and then rethrows it.

New internal assert failure looks like:

```
0 INTERNAL ASSERT FAILED at ../c10/test/util/exception_test.cpp:64, please report a bug to PyTorch.
Exception raised from TestBody at ../c10/test/util/exception_test.cpp:64 (most recent call first):
frame #0: <unknown function> + 0x6aab9 (0x7ff611d3aab9 in /data/users/ezyang/pytorch-tmp/build/lib/libc10.so)
frame #1: ...
```

Error message with context looks like:

```
This is an error
  This is context 1
  This is context 2
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21202891](https://our.internmc.facebook.com/intern/diff/D21202891)

[ghstack-poisoned]

ezyang committed:
Update on "Refactor error msg stack handling, add TORCH_RETHROW"


Fixes #36954.

The basic concept is to streamline the process of rethrowing
c10::Error with extra error information.  This is in a few
steps:

- I completely remodeled the Error data type and the internal
  invariants.  Instead of manually adding in newlines, the
  message stack formatting process is responsible for inserting
  newlines and spacing as necessary.  Call sites are then
  modified to respect the new API model.
- TORCH_RETHROW macro is added, which adds context to an error
  message and then rethrows it.

New internal assert failure looks like:

```
0 INTERNAL ASSERT FAILED at ../c10/test/util/exception_test.cpp:64, please report a bug to PyTorch.
Exception raised from TestBody at ../c10/test/util/exception_test.cpp:64 (most recent call first):
frame #0: <unknown function> + 0x6aab9 (0x7ff611d3aab9 in /data/users/ezyang/pytorch-tmp/build/lib/libc10.so)
frame #1: ...
```

Error message with context looks like:

```
This is an error
  This is context 1
  This is context 2
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21202891](https://our.internmc.facebook.com/intern/diff/D21202891)

[ghstack-poisoned]
